src/app/modules/item/services/threads.service.ts

```diff
@@ -44,37 +98,115 @@ interface TokenData {
   canWatchParticipant: boolean,
 }
 
-type Action = 'open-thread' | 'close-thread' | 'follow';
+type ThreadAction =
+  | { action: 'open-thread', history: ActivityLog[] }
+  | { action: 'close-thread' }
+  | { action: 'follow' }
+  | { action: 'unfollow' }
+  | { action: 'thread-status' }
+  | { action: 'send-message', message: string };
 
 @Injectable({
   providedIn: 'root',
 })
-export class ThreadService {
+export class ThreadService implements OnDestroy {
+
+  events$: Observable<ThreadEventMessage[]>;
+  status$: Observable<ThreadStatusMessage['status']>;
 
+  private newEvents$: Observable<ThreadEventMessage[]>;
+  private clearEvents$ = new Subject<void>();
   private tokenData?: TokenData;
   private socket: WebSocketSubject<unknown>;
-  messages$: Observable<ThreadEvent[]>;
+  private incomingMessages$: Observable<ThreadMessage[]>;
+  private subscriptions: Subscription[] = [];
 
-  constructor() {
+  constructor(
+    private activityLogService: ActivityLogService,
+  ) {
     if (!appConfig.forumServerUrl) throw new Error('cannot instantiate threads service when no forum server url');
     this.socket = webSocket(appConfig.forumServerUrl);
-    this.messages$ = this.socket.pipe(
-      decodeSnakeCase(D.array(threadEventDecoder)),
-      scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]),
+
+    this.incomingMessages$ = this.socket.pipe(
+      decodeSnakeCase(D.array(threadMessageDecoder)),
       catchError(() => EMPTY), // ignore undecoded messages
     );
+
+    this.newEvents$ = this.incomingMessages$.pipe(map(messages => messages.filter(isThreadEventMessage)));
+
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
+      map(messages => messages.sort((a, b) => a.time.valueOf() - b.time.valueOf())), // sort by date ascending
+    );
+
+    this.status$ = this.incomingMessages$.pipe(
+      map(messages => messages.find(isThreadStatusMessage)?.status),
+      filter(isNotUndefined),
+      distinctUntilChanged(),
+      shareReplay(1),
+    );
+
+    this.subscriptions.push(
+      // Every time we receive an event related to thread status, refetch the status.
+      this.newEvents$.pipe(
+        map(events => events.find(event => event.eventType === 'thread_opened' || event.eventType === 'thread_closed')),
+        filter(isNotUndefined),
+      ).subscribe(() => this.getStatus()),
```

Comment:
Because the handler "thread-status" was required in the spec of the backend, to me it means we want to code the feature on the backend instead of the frontend: https://github.com/France-ioi/AlgoreaForum/issues/20

Reply:
And so you consider we'll call the backend each time we receive a thread_opened or thread_closed to know whether the thread is closed or not?

Reply:
If we infer the thread status on the frontend from the received thread events (which is fine by me), then to me there's no point in creating a 'thread-status' service on any backend. Should I do that?